File: minimal/typescript/src/policy-bash.ts
import { spawn } from "node:child_process";
import type { Config } from "./config.js";

export type PolicyResult = "auto" | "ask" | "deny";

const BASH_TIMEOUT = 30000;

// Dangerous file patterns - deny reading these files
const DANGEROUS_FILE_PATTERNS: RegExp[] = [
  // Secret files
  /\.env$/,
  /\.env\./,
  /\.dev\.vars$/,
  /credentials/i,
  /secret/i,
  /\.pem$/,
  /\.key$/,
  /id_rsa/,
  /id_ed25519/,
  // Large/binary files
  /package-lock\.json$/,
  /yarn\.lock$/,
  /pnpm-lock\.yaml$/,
  /\.DS_Store$/,
  /node_modules/,
];

// Destructive command patterns - always deny
const BUILTIN_DENY_PATTERNS: RegExp[] = [
  /rm\s+(-[rf]+\s+)*\//,
  /rm\s+-rf?\s+\*/,
  /rm\s+-rf?\s+\.\*/,
  /mkfs/,
  /dd\s+if=.*of=\/dev/,
  />\s*\/dev\/sd/,
  /gcloud\s+.*delete/,
  /gcloud\s+.*destroy/,
  /aws\s+.*delete/,
  /aws\s+.*terminate/,
  /kubectl\s+delete/,
  /:\(\)\s*\{.*\|.*&.*\}/,
  /chmod\s+-R\s+777\s+\//,
  /chown\s+-R.*\//,
  /curl.*\|\s*(ba)?sh/,
  /wget.*\|\s*(ba)?sh/,
];

const BUILTIN_AUTO_COMMANDS = [
  "ls",
  "pwd",
  "whoami",
  "date",
  "which",
  "cat",
  "head",
  "tail",
  "less",
  "more",
  "wc",
  "file",
  "stat",
  "tree",
  "find",
  "fd",
  "grep",
  "rg",
  "git status",
  "git diff",
  "git log",
  "git branch",
];

const FORCE_ASK_PATTERN = /[|;&`$()]/;

export function checkPolicy(command: string, config: Config): PolicyResult {
  const cmd = command.trim();

  // Build deny patterns
  const denyPatterns = [
    ...BUILTIN_DENY_PATTERNS,
    ...config.policy.denyPatterns.map((p) => new RegExp(p)),
  ];

  // Build auto commands
  const autoCommands = [...BUILTIN_AUTO_COMMANDS, ...config.policy.autoCommands];

  // 1. Check destructive command patterns
  for (const pattern of denyPatterns) {
    if (pattern.test(cmd)) {
      return "deny";
    }
  }

  // 2. Check dangerous file patterns in args
  const args = cmd.split(/\s+/).slice(1).join(" ");
  for (const pattern of DANGEROUS_FILE_PATTERNS) {
    if (pattern.test(args)) {
      return "deny";
    }
  }

  // 3. Force ask for complex commands
  if (FORCE_ASK_PATTERN.test(cmd)) {
    return "ask";
  }

  // 4. Check auto
  for (const autoCmd of autoCommands) {
    if (cmd === autoCmd || cmd.startsWith(autoCmd + " ")) {
      return "auto";
    }
  }

  // 5. Default action
  return config.policy.defaultAction;
}

export interface BashResult {
  stdout: string;
  stderr: string;
  code: number;
}

export function formatCommandResult(command: string, result: BashResult): string {
  let content = `[command] ${command}`;
  if (result.stdout) content += `\n[stdout]\n${result.stdout.trimEnd()}`;
  if (result.stderr) content += `\n[stderr]\n${result.stderr.trimEnd()}`;
  if (result.code !== 0) content += `\n[exit_code] ${result.code}`;
  return content;
}

export function runBash(command: string, workspaceRoot: string): Promise<BashResult> {
  return new Promise((resolve) => {
    const child = spawn(command, {
      shell: true,
      cwd: workspaceRoot,
      env: process.env,
    });

    let stdout = "";
    let stderr = "";
    let timedOut = false;

    const timeout = setTimeout(() => {
      timedOut = true;
      child.kill("SIGTERM");
    }, BASH_TIMEOUT);

    child.stdout.on("data", (data) => {
      stdout += String(data);
    });
    child.stderr.on("data", (data) => {
      stderr += String(data);
    });
    child.on("close", (code) => {
      clearTimeout(timeout);
      if (timedOut) {
        resolve({ stdout, stderr: "Command timed out (30s)", code: 124 });
      } else {
        resolve({ stdout, stderr, code: code ?? 0 });
      }
    });
  });
}

File: minimal/rust/src/policy_bash.rs
use crate::config::Config;
use regex::Regex;
use std::process::Command;
use std::time::Duration;
use wait_timeout::ChildExt;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PolicyResult {
    Auto,
    Ask,
    Deny,
}

const BASH_TIMEOUT: Duration = Duration::from_secs(30);

const DANGEROUS_FILE_PATTERNS: &[&str] = &[
    r"\.env$",
    r"\.env\.",
    r"\.dev\.vars$",
    r"credentials",
    r"secret",
    r"\.pem$",
    r"\.key$",
    r"id_rsa",
    r"id_ed25519",
    r"package-lock\.json$",
    r"yarn\.lock$",
    r"pnpm-lock\.yaml$",
    r"\.DS_Store$",
    r"node_modules",
];

const BUILTIN_DENY_PATTERNS: &[&str] = &[
    r"rm\s+(-[rf]+\s+)*\/",
    r"rm\s+-rf?\s+\*",
    r"rm\s+-rf?\s+\.\*",
    r"mkfs",
    r"dd\s+if=.*of=\/dev",
    r">\s*\/dev\/sd",
    r"gcloud\s+.*delete",
    r"gcloud\s+.*destroy",
    r"aws\s+.*delete",
    r"aws\s+.*terminate",
    r"kubectl\s+delete",
    r":\(\)\s*\{.*\|.*&.*\}",
    r"chmod\s+-R\s+777\s+\/",
    r"chown\s+-R.*\/",
    r"curl.*\|\s*(ba)?sh",
    r"wget.*\|\s*(ba)?sh",
    r"ls\s+-[^\s]*R",
    r"ls\s+-R",
    r"sed\s.*-i",
];

const BUILTIN_AUTO_COMMANDS: &[&str] = &[
    "ls",
    "ls -la",
    "ls -l",
    "ls -a",
    "pwd",
    "whoami",
    "date",
    "which",
    "cat",
    "head",
    "tail",
    "less",
    "more",
    "wc",
    "file",
    "stat",
    "tree",
    "find",
    "fd",
    "grep",
    "rg",
    "sed -n",
    "git status",
    "git diff",
    "git log",
    "git branch",
];

pub fn check_policy(command: &str, config: &Config) -> PolicyResult {
    let cmd = command.trim();

    let mut deny_patterns: Vec<Regex> = BUILTIN_DENY_PATTERNS
        .iter()
        .filter_map(|pattern| Regex::new(pattern).ok())
        .collect();
    for pattern in &config.policy.deny_patterns {
        if let Ok(re) = Regex::new(pattern) {
            deny_patterns.push(re);
        }
    }

    for pattern in deny_patterns {
        if pattern.is_match(cmd) {
            return PolicyResult::Deny;
        }
    }

    let args = cmd
        .split_whitespace()
        .skip(1)
        .collect::<Vec<_>>()
        .join(" ");
    for pattern in DANGEROUS_FILE_PATTERNS {
        if let Ok(re) = Regex::new(pattern) {
            if re.is_match(&args) {
                return PolicyResult::Deny;
            }
        }
    }

    let force_ask = Regex::new(r"[|;&`$()]").unwrap();
    if force_ask.is_match(cmd) {
        return PolicyResult::Ask;
    }

    let mut auto_commands: Vec<String> = BUILTIN_AUTO_COMMANDS
        .iter()
        .map(|value| value.to_string())
        .collect();
    auto_commands.extend(config.policy.auto_commands.iter().cloned());

    for auto_cmd in auto_commands {
        if cmd == auto_cmd || cmd.starts_with(&(auto_cmd + " ")) {
            return PolicyResult::Auto;
        }
    }

    if config.policy.default_action == "deny" {
        PolicyResult::Deny
    } else {
        PolicyResult::Ask
    }
}

#[derive(Clone, Debug)]
pub struct BashResult {
    pub stdout: String,
    pub stderr: String,
    pub code: i32,
}

pub fn format_command_result(command: &str, result: &BashResult) -> String {
    let mut content = format!("[command] {}", command);
    if !result.stdout.trim().is_empty() {
        content.push_str("\n[stdout]\n");
        content.push_str(result.stdout.trim_end());
    }
    if !result.stderr.trim().is_empty() {
        content.push_str("\n[stderr]\n");
        content.push_str(result.stderr.trim_end());
    }
    if result.code != 0 {
        content.push_str(&format!("\n[exit_code] {}", result.code));
    }
    content
}

pub fn run_bash(command: &str, workspace_root: &str) -> BashResult {
    let mut cmd = Command::new("bash");
    cmd.arg("-c")
        .arg(command)
        .current_dir(workspace_root)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped());

    let mut child = match cmd.spawn() {
        Ok(child) => child,
        Err(err) => {
            return BashResult {
                stdout: String::new(),
                stderr: format!("Failed to spawn bash: {}", err),
                code: 1,
            };
        }
    };

    let mut stdout = child.stdout.take();
    let mut stderr = child.stderr.take();

    let stdout_handle = std::thread::spawn(move || {
        let mut buf = Vec::new();
        if let Some(mut out) = stdout.take() {
            let _ = std::io::Read::read_to_end(&mut out, &mut buf);
        }
        buf
    });

    let stderr_handle = std::thread::spawn(move || {
        let mut buf = Vec::new();
        if let Some(mut err) = stderr.take() {
            let _ = std::io::Read::read_to_end(&mut err, &mut buf);
        }
        buf
    });

    let mut timed_out = false;
    let code = match child.wait_timeout(BASH_TIMEOUT) {
        Ok(Some(status)) => status.code().unwrap_or(1),
        Ok(None) => {
            timed_out = true;
            let _ = child.kill();
            let _ = child.wait();
            124
        }
        Err(_) => 1,
    };

    let stdout = stdout_handle.join().unwrap_or_default();
    let stderr = stderr_handle.join().unwrap_or_default();

    let stdout = String::from_utf8_lossy(&stdout).to_string();
    let stderr = String::from_utf8_lossy(&stderr).to_string();

    if timed_out {
        return BashResult {
            stdout,
            stderr: "Command timed out (30s)".to_string(),
            code,
        };
    }

    BashResult { stdout, stderr, code }
}
